packages/ui/src/features/data-points/useDatapoints.ts

```diff
@@ -0,0 +1,230 @@
+import {
+  ValueVariableType,
+  CategoricalVariableType,
+  ScoreVariableType,
+  DataEntry,
+  CombinedVariableType,
+} from '@boostv/process-optimizer-frontend-core'
+import { useCallback, useMemo } from 'react'
+import { TableDataRow } from '../core'
+import produce from 'immer'
+
+export const useDatapoints = (
+  valueVariables: ValueVariableType[],
+  categoricalVariables: CategoricalVariableType[],
+  scoreVariables: ScoreVariableType[],
+  dataPoints: DataEntry[]
+) => {
+  const scoreNames = useMemo(
+    () => scoreVariables.filter(it => it.enabled).map(it => it.name),
+    [scoreVariables]
+  )
+
+  const state = useMemo(
+    () =>
+      ({
+        rows: buildRows(
+          valueVariables,
+          categoricalVariables,
+          scoreNames,
+          dataPoints
+        ),
+        meta: dataPoints.map(dp => dp.meta),
+      } as const),
+    [categoricalVariables, dataPoints, scoreNames, valueVariables]
+  )
+
+  const addRow = useCallback(
+    (row: TableDataRow) =>
+      _addRow(
+        dataPoints,
+        convertToDataEntry(
+          valueVariables,
+          categoricalVariables,
+          scoreVariables,
+          row
+        )
+      ),
+    [categoricalVariables, dataPoints, scoreVariables, valueVariables]
+  )
+
+  const deleteRow = useCallback(
+    (rowIndex: number) => _deleteRow(dataPoints, rowIndex),
+    [dataPoints]
+  )
+  const editRow = useCallback(
+    (rowIndex: number, row: TableDataRow) =>
+      _editRow(dataPoints, rowIndex, row),
+    [dataPoints]
+  )
+  const setEnabledState = useCallback(
+    (rowIndex: number, enabled: boolean) =>
+      _setEnabledState(dataPoints, rowIndex, enabled),
+    [dataPoints]
+  )
+
+  return {
+    state,
+    addRow,
+    deleteRow,
+    editRow,
+    setEnabledState,
+  }
+}
+
+const convertToDataEntry = (
+  valueVariables: ValueVariableType[],
+  categoricalVariables: CategoricalVariableType[],
+  scoreVariables: ScoreVariableType[],
+  row: TableDataRow
+) => {
+  const data = row.dataPoints.map(dp => ({
+    name: dp.name,
+    value:
+      categoricalVariables.find(cv => cv.name === dp.name) !== undefined
+        ? String(dp.value ?? '')
+        : Number(dp.value ?? 0),
+  })) satisfies DataEntry['data']
+  const meta = {
+    enabled: row.enabled ?? true,
+    id: row.metaId ?? 0,
+    valid: row.valid ?? true,
+  } satisfies DataEntry['meta']
+  if (
+    data.length <
+      valueVariables.length +
+        categoricalVariables.length +
+        scoreVariables.length -
+        1 &&
+    scoreVariables[1] !== undefined
+  ) {
+    data.push({ name: scoreVariables[1].name, value: 0 })
+  }
+  return { data, meta } satisfies DataEntry
+}
+
+const _addRow = (original: DataEntry[], newRow: DataEntry) =>
```

Comment:
Isn't this `const metaId = Math.max(0, ...state.meta.map(m => m.id)) + 1` from the reducer's `rowAdded` (line 75) missing here?